ci2/ci/github.py

```diff
@@ -128,17 +165,78 @@ def from_gh_json(gh_json, target_branch):
                 review_state = 'approved'
             else:
-                assert state == 'DISMISSED' or state == 'COMMENTED', state
+                assert state in ('DISMISSED', 'COMMENTED'), state
 
         self.review_state = review_state
 
+    async def start_build(self, batch_client):
+        assert not self.batch
+
+        # FIXME this needs to be per-PR and async
+        try:
+            async with repos_lock:
+                target_repo = self.target_branch.branch.repo
+                repo_dir = f'repos/{target_repo.short_str()}'
+
+                shell('git', 'config', 'user.email', 'hail-ci-leader@example.com')
+                shell('git', 'config', 'user.name', 'hail-ci-leader')
+
+                if not os.path.isdir(repo_dir):
+                    os.makedirs(repo_dir, exist_ok=True)
+                    shell('git', '-C', repo_dir, 'clone', self.target_branch.branch.repo.url, '.')
+
+                if sp.run([
+                        '/bin/sh', '-c',
+                        f'git -C {shq(repo_dir)} remote | grep -q {shq(self.source_repo.short_str())}'
+                ]).returncode != 0:
+                    shell('git', '-C', repo_dir, 'remote', 'add', self.source_repo.short_str(), self.source_repo.url)
+
+                shell('git', '-C', repo_dir, 'reset', '--merge')
+                shell('git', '-C', repo_dir, 'fetch', 'origin')
+                shell('git', '-C', repo_dir, 'fetch', self.source_repo.short_str())
+                shell('git', '-C', repo_dir, 'checkout', self.target_branch.sha)
+                shell('git', '-C', repo_dir, 'merge', self.source_sha, '-m', 'merge PR')
+
+                self.sha = (sp.check_output(['git', 'rev-parse', 'HEAD'])
+                            .decode('utf-8')
+                            .strip())
+
+                with open(f'{repo_dir}/build.yaml', 'r') as f:
+                    config = BuildConfiguration(self, f.read())
+        except (sp.CalledProcessError, FileNotFoundError) as e:
+            log.exception(f'could not get hail-ci-build-image due to {e}')
+            self.build_state = 'merge_failure'
+            return
+
+        batch = None
+        try:
+            batch = await batch_client.create_batch(
+                attributes={
+                    'token': secrets.token_hex(16),
+                    'target_branch': self.target_branch.branch.short_str(),
+                    'pr': str(self.number),
+                    'source_sha': self.source_sha,
+                    'target_sha': self.target_branch.sha
+                })
+            await config.build(batch, self)
+            await batch.close()
+            self.batch = batch
+        finally:
+            if batch and not self.batch:
+                batch.cancel()
+
     async def heal(self, batch, run, seen_batch_ids):
+        if self.build_state is not None:
+            if self.batch:
+                seen_batch_ids.add(self.batch.id)
+            return
```

Comment:
build_state seems to take on four values: merge_failure, success, failure, None. In all three non-None cases, batch should be None, right?

Reply:
I don't clear the batch after it finishes, so success and failure should self.batch the pipeline that exhibited the success or failure.